Add toolbar modes

eranl opened the pull request:
This is the resurrected #838 by @codokie, brought up to current code, with a few other changes.
This change leaves the clipboard and emoji toolbars intact and functional when the main toolbar is hidden, resulting in a smaller net height. Is this the desired effect?

Fixes #374, fixes #1153, fixes #1379, fixes #961, fixes #732.

Helium314 commented:
> This change leaves the clipboard and emoji toolbars intact and functional when the main toolbar is hidden, resulting in a smaller net height. Is this the desired effect?

Not sure whether I understand: the toolbar is shown in emoji and clipboard view, but the total height is the same as the main keyboard? (I didn't yet test or check the code)

eranl commented:
Exactly.

Helium314 commented:
I guess some users might complain, but also some users _did_ complain in very old versions because there was a height change exactly because of the toolbar. So I'd say the current state is preferable.
Would it be reasonably simple to have a setting for this? (no need to add it yet)

eranl commented:
> > add a setting for choosing between the current state of this PR and this third option?

Done.

Helium314 commented on app/src/main/java/helium314/keyboard/keyboard/KeyboardSwitcher.java:
```diff
@@ -539,6 +545,11 @@ public void showToast(final String text, final boolean briefToast){
     }
 
+    private static int getSecondaryStripVisibility() {
+        return Settings.getValues().mToolbarMode == ToolbarMode.HIDDEN && Settings.getValues().mToolbarHidingGlobal?
```
You always query `mToolbarHidingGlobal` together with checking whether `mToolbarMode` is `HIDDEN`. I think it would be easier and more understandable if you used some `mSecondaryStripVisible` (choose a name) in SettingsValues instead.

eranl replied:
Done.

eranl commented:
> I don't really understand some of the changes in `LatinIME`:

I must admit that I didn't follow all of them either... Initially, I only removed the changes that were obviously unrelated.

> * Why the different way of setting RTL mode? It doesn't seem related to toolbar modes.

Reverted.

> * Why is `SuggestionStripView.updateKeys` now sometimes handled by `LatinIME`?

Reverted.

> * Why is the toolbar hiding in `hideWindow` removed?

Otherwise, it breaks in toolbar-only mode.

> * Why actually set a non-null `mSuggestionStripView` when the toolbar is hidden? Then also some of the settings checks can be removed because `hasSuggestionStripView` is already used. Slight adjustments needed for computing insets and inline suggestions. 

Done.

> Btw when not showing suggestions, `onCreateInlineSuggestionsRequest` should probably return null to clarify that inline suggestions are not supported.

Done. How do I test this?

eranl commented on app/src/main/java/helium314/keyboard/latin/suggestions/SuggestionStripView.java:
```diff
@@ -156,6 +158,13 @@ public SuggestionStripView(final Context context, final AttributeSet attrs, fina
+            setToolbarVisibility(true);
+        } else if (mToolbarMode == ToolbarMode.SUGGESTION_STRIP){
+            setToolbarVisibility(false);
+        }
```
Reverted.

Helium314 commented:
>> * Why actually set a non-null `mSuggestionStripView` when the toolbar is hidden? Then also some of the settings checks can be removed because `hasSuggestionStripView` is already used. Slight adjustments needed for computing insets and inline suggestions. 
>
> Done.

What I meant was just setting a null `mSuggestionStripView` in `setInputView` if suggestion strip is switched off. Did you try that?
Interacting with the strip when it's not even displayed feels like a waste of resources, and the `hasSuggestionStripView` is already in place almost everywhere it's needed.
In a short test it seems to work for me (after adding another check in `onComputeInsets`).

Helium314 commented:
> > What I meant was just setting a null `mSuggestionStripView` in `setInputView` if suggestion strip is switched off.
> 
> Done.

Looks like it doesn't really work, `mSuggestionStripView` is still not null when hidden. Maybe it's a timing issue, i.e. visibility is set after `setInputView`.
With something like `mSuggestionStripView = mSettings.getCurrent().mToolbarMode == ToolbarMode.HIDDEN ? null : view.findViewById(R.id.suggestion_strip_view);` it's actually null.
(It probably does not really matter whether it's really null, but it just doesn't feel right with all the `hasSuggestionStripView()` checks)

eranl commented:
> Looks like it doesn't really work

Oops, thanks. Fixed.
